dev-tools/mage/gotool/modules.go

```diff
@@ -0,0 +1,43 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package gotool
+
+type goMod func(opts ...ArgOpt) error
+
+// Mod is the command go mod.
+var Mod goMod = runGoMod
+
+// Tidy cleans the go.mod file
+func (goMod) Tidy() error {
+	return runGoMod(posArg("tidy"))
+}
+
+// Vendor downloads and copies dependencies under the folder vendor.
+func (goMod) Vendor() error {
+	return runGoMod(posArg("vendor"))
+}
+
+// Verify check that deps have the expected content.
+func (goMod) Verify() error {
+	return runGoMod(posArg("verify"))
+}
+
+func runGoMod(opts ...ArgOpt) error {
+	args := buildArgs(opts)
+	return runVGo("mod", args)
+}
```

Comment:
Be careful with `posArg`. If one adds other flags in the future, then the commands become `go mod <flags> <sub-command>`. But the commands require you to run `go mod <sub-command> <flags>`.

In magetools I work around this limitation for `go run` by wrapping ArgOpts that get translated into positional args (See: [run.go](https://github.com/urso/magetools/blob/master/gotool/run.go#L11)) . But `go run` can be called like this: `go run <build flags> <list of go files> <args passed to main>`. It is used [like this](https://github.com/urso/diag-ecs/blob/master/magefile.go#L116).

`go mod` is simpler fortunately, but let's use a pattern like this:

```
type goMod struct {
  Tidy ModTidy
  Verify ModVerify
  Vendor ModVendor
}

type ModTidy func(opts ...ArgOpt) error
type ModVerify func(opts ...ArgOpt) error
type ModVendor func(opts ...ArgOpt) error

type modCommand struct {
  method string
}

var Mod = goMod {
  Tidy = modCommand{"tidy"}.run,
  Verify = modCommand{"verify"}.run,
  Vendor = modCommand{"vendor"}.run,
}

func (cmd modCommand) run(args ...ArgOpt) error {
  ...  
}
```

Using this structure we can define flags on `goMod`, `ModTidy`, and others. E.g. `tidy`, and `vendor` support a verbose flag `-v`, but `verify` does not.

```
func (ModTidy) Verbose(b bool) argOpt { return BoolFlag("-v", b) }

func (ModVendor) Verbose(b bool) argOpt { return BoolFlag("-v", b) }
```

When updating the go version we will also need to do some maintenance. `go mod edit` might be nice to have as well in the future

```
func (ModEdit) SetGoVersion(version string) argOpt { ... }
func (ModEdit) PrintJSON(b bool) argOpt { ... }
func (ModEdit) Format(b bool) argOpt { ... }
```

Then one can use:

```
goMod := gotool.Mod
goMod.Edit(
  goMod.Edit.SetGoVersion(version),
  goMod.Edit.Format(true),
)
```